박은지 과제 제출합니다.

eunjipark123 opened the pull request:
1차 제출입니다.

eunjipark123 committed:
과제 제출

1차 제출

eunjipark123 committed:
파일 경로 수정

Moon-edu commented on tests/homework/csv_hw.py:
```diff
@@ -4,19 +4,74 @@
+            if d[4] == max(assets) :
+                results = ((d[0], d[4]))
+
+    return print(results)
```
결과를 print하는 것이 아니고 결과 값을 리턴해야합니다.
참고로 print함수의 리턴 타입은 `None`입니다

Moon-edu replied:
그리고 여기 함수 선언을 보시면 Tuple타입이 리턴됨을 알 수 있습니다.
```
def find_richest_and_asset() -> Tuple:
```
하지만 이 로직은 list를 리턴하고 있습니다.

Moon-edu replied:
함수 위의 주석을 보고 리턴 값이 어떤지 다시 한 번 고민해보시기 바랍니다.
```
# 아래 함수를 실행하면 Tuple 값이 리턴되며, 자산 평가액이 가장 많은 사람의 이름과 자산 평가액이 리턴되어야 합니다.
# 리턴 값의 예) ("Donald", 1023014)
```

eunjipark123 replied:
변경했습니다 
- tuple로 리턴되도록 설정 
- tuple의 두 번째 값이 int로 보이도록 설정

Moon-edu commented on tests/homework/csv_hw.py:
```diff
@@ -4,19 +4,74 @@
 # 리턴 값의 예) ["Dublin", "Seoul", "New York"]
 def find_top3_richest_city() -> list:
-    pass
+    with open("hw_data/assets.csv","r") as csv_file :
```
Python의 Coding style guide를 보면, colon은 띄어쓰기를 하지 않습니다.
reference: https://peps.python.org/pep-0008/ 
![image](https://github.com/Moon-edu/de-course-week4/assets/102866148/463d32ed-4157-4bad-8c8e-ccb1ccab8820)

eunjipark123 replied:
띄어쓰기 수정했습니다.

Moon-edu commented on tests/homework/csv_hw.py:
```diff
@@ -4,19 +4,74 @@
+                for row in city_assets :
+                    if d[2] == row[0] :
+                        row[1] = int(row[1]) + int(d[4])
+                        row[2] = int(row[2]) + 1
```
시간/공간 복잡도가 좀 많이 안 좋은 것 같습니다.
for 루프에서 매번 set을 생성하며, not in 연산을 list까지 하고 있습니다. 대략적으로 계산해보자면
필요한 공간과 city_assets 순회 비용이 1+2+3+ .... n이라서 n^2.

전반적으로 코드가 이해하기 어려운 알고리즘이고 복잡도면에서 개선이 많이 필요한 로직으로 보입니다.

어떠한 알고리즘을 구현할 때, 고민해야할 것은 어떠한 자료구조를 써야할 것인가입니다. 
저라면 이런 문제를 마주 쳤을 때 이런 과정으로 고민할 것 같습니다.
데이터를 보면 이렇게 되어 있는데
a: 100
b: 200
c: 300
a: 100
b: 500

1. 도시별 자산 평가액의 평균을 내니깐, 도시별로 묶어보자
a: 100, 100
b: 200, 500
c: 300
2. 묶으려면 어떤 자료구조가 좋을까? 나는 도시 이름도 필요하고 각각의 자산 평가액도 필요해. dictionary에다가 도시 이름을 key로 값을 자산 평가액 리스트로 하면 같은 도시의 자산 평가액이 한 곳으로 묶이겠지

a: (100 + 100) / 2 = 100
b: (200 + 500) / 2 = 350
c: 300 / 1  = 300
3. 그 다음에 각각의 도시에 대해 묶여있는 자산 평가액을 다 더한 후 자산평가액의 갯수로 나누면 평균을 얻을 수 있을거야

b: 350
c: 300
a: 100
4. 그 다음에 자산 평가액으로 정렬

b,c,a
6.  상위 3개에서 도시 이름만 가져오면 끝

아직 코딩이 익숙하지 않아 그런 것일 수 있고, 제가 코딩을 많이 해봐서 쉽게 말하는 것일 수도 있습니다. 계속 이러한 문제를 풀다보면 코딩에 익숙해지게 되고 자연스럽게 이런 과정으로 생각하게 될겁니다. 결론은 코딩은 많이 해봐야한다는 것입니다. 😄

eunjipark123 replied:
말씀해주신 로직으로 변경해봤습니다..!

Moon-edu commented on tests/homework/excel_hw.py:
```diff
@@ -4,16 +4,69 @@
+            'city' : row[2].value,
+            'age' : int(row[3].value),
+            'est_asset_dollar' : int(row[4].value)
+        })
```
잘 하셨습니다. 다만 아쉬운 점이 있다면, 결과를 내기 위해 필요한 데이터는 회사명과 자산 평가액 두 개입니다. 나머지 데이터는 필요없기때문에 메모리 낭비가 됩니다.

eunjipark123 replied:
변경했습니다. 메모리도 고려해야 한다는 점을 꼭 기억하겠습니다 :)

Moon-edu commented on tests/homework/xml_hw.py:
```diff
@@ -4,19 +4,62 @@
+            if int(row['est_asset_dollar']) == min(asset) :
+                min_asset.append((row['name'], row['est_asset_dollar']))
+
+    return(print(min_asset[0]))
```
박은지 님께만 드리는 퀴즈입니다.
여기 이 함수의 시간복잡도와 공간복잡도를 생각해보고 답글을 달아주세요 😀

eunjipark123 replied:
평균을 구할 때와 하위 3개를 구할 때 모두 반복문을 사용해서 시간, 공간 복잡도가 높을 것으로 예상됩니다.  특히 하위3개 구할 때 while문을 반복해서 실행하기 때문에 시간 복잡도가 훨씬 높았을 것 같아요.

eunjipark123 committed:
과제 수정

1. 리턴값(데이터타입) 변경
2. colon 띄어쓰기 수정
3. top3, bottom3 로직 변경 (반복문 최소화)
4. xlsx 파일에서 필요한 데이터만 가지고 오기

eunjipark123 committed:
부분 수정

indent 수정

eunjipark123 committed:
부분 수정

리스트 괄호 수정

eunjipark123 committed:
수정제출

컬럼 위치 수정 (Name -> City)

Moon-edu commented on tests/homework/csv_hw.py:
```diff
@@ -4,19 +4,51 @@
+        data = [(i[0], int(i[4])) for i in csv_data]
+
+        # data에서 asset값이 최대인 경우만 출력
+        max_asset = max(data, key=lambda x: x[1]) # 최대값을 표현할 수 있음
```
현재 이 시점에서는 이미 파일을 모두 읽어온 상황이니 파일 스트림을 계속 열고 있을 필요가 없겠죠?  
코드의 위치로 보면 이 코드가 실행되는 시점에도 파일을 연 채로 그대로 두고 있습니다.  
리소스 낭비죠

Moon-edu replied:
이러한 패턴이 다른 파일에서도 보이고 있습니다.

eunjipark123 replied:
오류 없는 선에서 인덴트 수정했습니다 :)

Moon-edu commented on tests/homework/excel_hw.py:
```diff
@@ -4,16 +4,61 @@
+        })
+
+    # corp.로 끝나는 회사명 가지고 오기 -> 스페이스로 회사명을 split하고 마지막 단위가 Corp이면 가져오기
+    corp_company = [row['est_asset_dollar'] for row in data if row['company'].split()[-1] == 'Corp.']
```
결과적으로 정답이 맞을지는 몰라도 제시된 문제의 조건에는 부합하지 않는 로직입니다.
"Corp.로 끝나는"이 조건이기때문에 Space가 없는 ACorp.인 회사가 있었다면 이 로직은 오답을 도출했을겁니다.

eunjipark123 replied:
아하 문제 의도를 다르게 이해했습니다. split() -> endswith()로 로직 변경해봤습니다!

Moon-edu commented on tests/homework/excel_hw.py:
```diff
@@ -4,16 +4,61 @@
+        })
+
+    # 평균 구하기
+    llc_company = [row['est_asset_dollar'] for row in data if row['company'].split()[-1] == 'LLC']
```
현재 데이터를 기준으로는 맞지만, 여기서도 효율성의 문제를 따진다면 최적화된 코드는 아닙니다.
항상 코드를 작성할 때는 시간복잡도, 공간복잡도가 이게 최선인가를 생각해보셔야합니다.
평균을 계산하는데는 시간 복잡도는 O(n) 이며, 1번의 파일 읽기로 가능합니다.
공간 복잡도는 O(1)이며 두 개의 변수면 충분합니다. sum과 count입니다.   
이 로직의 시간/공간복잡도가 어떻게 되는지 생각해보시구요.
두 번째로 로직 역시 잘못된 로직입니다. Corp.의 케이스와 똑같은 이유로 오답입니다.

eunjipark123 replied:
파일 가져올 때 조건에 해당하는 회사의 asset만 가져오도록 변경해봤습니다. 반복문을 파일 가져올 때 한 번으로 수정했는데, 제가 이해한 내용이 맞을까요~?

Moon-edu replied:
```
for row in ws:
        if not wb_flag:
            wb_flag = True
            continue
        data.append({
            'company': row[1].value,
            'est_asset_dollar': int(row[4].value)
        })
```
파일을 가져올 때 조건에 해당하는 회사의 asset만 가져온 것 같지 않습니다.
제가 코멘트한 부분은 사실 파일을 가져올때부터 추가적인 메모리(`data`)를 할당할 필요가 없다는 겁니다. 평균은 전체 합 / 갯수입니다. 그래서 필요한 것은 llc로 끝나는 회사의 자산만 더한 결과와 총 갯수 이 두 변수면 충분하다는 것입니다.
그리고 sum이라는 함수를 호출하는데 이 역시 꽤 비싼 연산입니다. 리스트를 전체 순회하면서 값을 더하니깐요.

eunjipark123 replied:
아항 다시 해보겠습니다!

Moon-edu commented on tests/homework/json_hw.py:
```diff
@@ -4,9 +4,22 @@
+    return (sum_asset, avg_asset, cnt_asset)
+
+result = summarize()
+print(result)
```
이 로직은 최종 제출에서는 필요없는 로직입니다.

eunjipark123 replied:
반영했습니다 :)

eunjipark123 committed:
수정제출

1. 인덴트 수정 : 파일 스트림 관련 인덴트 수정
2. 로직 변경 : split() -> endswith()로 변경
3. 복잡도 개선 : LLC로 끝날 때만 값을 불러와서 평균을 도출할 수 있게 변경했습니다. (하단 list comprehension은 삭제)
4. 오타 삭제 : 테스트 하려고 적어둔 건데 끼어있었습니다. 삭제했습니다 :)

eunjipark123 committed:
수정 제출

avgs 변수에서 len(data) = 0인 가능성 고려하기

eunjipark123 commented:
1. 이전에 제출한 함수에서 틀린 이유를 아직 잘 모르겠습니다. sum, len 값은 채점표와 동일하게 나왔거든요! 
sum=175985567, len=34, avg=5176046.088235294였습니다. 
abs(result - (175985567 / 34)) < 0.00001 이것도 True로 나왔는데.. 문제가 무엇이었을까요? 

2. 일단 sum, cnt 변수에 바로 값을 반영하도록 조정해봤습니다!

eunjipark123 committed:
로직 수정

data에 저장하지 않고 바로 sum, cnt 변수에 넣었습니다~

Moon-edu commented:
로직이 틀리진 않았습니다. 저는 시간/공간 복잡도를 계산해봤을때 비효율적이라는 이야기였습니다.
전체적으로 정답을 도출하는 로직은 맞습니다만, 최적화된 코드는 아니었다라는 이야기였습니다.  
최대값, 최소값을 찾는 로직에서, 불필요하게 먼저 리스트로 전부 모아온 후 최대값 최소값을 구하고 있다던지  
sum을 구하기 위해 먼저 리스트로 모아온 후 다시 sum을 호출하여 리스트를 한 번 더 스캔한다던지와 같은 비효율이 그 예입니다.

Moon-edu commented:
https://github.com/eunjipark123/de-course-week4/blob/homework/eunji_week4/tests/homework/csv_hw.py#L20C9-L20C9
이 코드를 보시면 최대값을 구하기 위해 굳이 공간복잡도 O(n)을 추가로 할당하고 있습니다.  
최대값을 구하는 로직은 현재값과 현재까지 최대값을 유지하는 변수만 있으면 됩니다.

eunjipark123 commented:
아핫 이해했습니다. 로직을 여러 번 생각해보면서 효율적인 방법을 찾아봐야겠네요!!  
제가 여러 번 헤메서 많이 답답하시죠 🥲🥲 
그래도 계속 봐주시고 좋은 방향으로 나아갈 수 있게 피드백 해주셔서 정말 감사드립니다! ㅎㅎ

Moon-edu commented on tests/homework/xml_hw.py:
```diff
@@ -4,19 +4,47 @@
+        data = xmltodict.parse(xml_file.read())['data']['asset']
+
+        # 자산평가액 str->int로 변경해 출력
+        data = [(row['name'], int(row['est_asset_dollar'])) for row in data]
```
여기도 마찬가지로 파일을 읽고나서 닫지않고 추가 로직을 실행하는 구문입니다. 이런 안티패턴이 여기뿐만 아니라 꽤나 보입니다.

eunjipark123 replied:
인덴트 오류 발생하지 않는 선에서 최대한 밖으로 끌어낸 것 같은데, 이슈가 있나보네요. 또 어떤 부분이 잘못되었을까요..?

Moon-edu replied:
xmltodict.parse가 완료된 시점에서 더 이상 파일을 열고 있을 필요는 없습니다.
```
with open("hw_data/assets.xml", "r") as xml_file:
        data = xmltodict.parse(xml_file.read())['data']['asset']

        # 자산평가액 str->int로 변경해 출력
        data = [(row['name'], int(row['est_asset_dollar'])) for row in data] # <-- 이 부분은 데이터를 전부 읽어온 상황에서도 with 구문내에 위치하고 있습니다. with 구문 안에서는 파일을 닫지 않은 상황인거죠, 하지만 파일이 열려있을 필요가 없는 상황입니다.

        # 최소값 구하기
        min_asset = min(data, key=lambda x: x[1]) 
```

eunjipark123 committed:
수정제출

1. 리스트에 담지 않고 바로 집계값(max, min, sum, cnt, avg) 출력
2. sum, len 함수 안 쓰기
3. 인덴트 수정

eunjipark123 committed:
부분수정

int추가

eunjipark123 commented on tests/homework/excel_hw.py:
```diff
@@ -14,23 +14,16 @@ def find_corp_total_asset() -> int:
-    # corp.로 끝나는 회사명 가지고 오기 -> 스페이스로 회사명을 split하고 마지막 단위가 Corp이면 가져오기
-    corp_company = [row['est_asset_dollar'] for row in data if row['company'].endswith('Corp.')]
-    sum_asset = sum(corp_company)
+        elif row[1].value.endswith('Corp.'):
```
1. 데이터 가져올 때 endswith 조건 사용하도록 변경 
2. sum함수 없이 value만 더해 sum 도출

eunjipark123 commented on tests/homework/csv_hw.py:
```diff
@@ -42,13 +37,15 @@ def find_top3_richest_city() -> list:
+                city_assets[city][1] += 1
 
-    avg_assets = {city:sum(asset)/len(asset) for (city, asset) in city_assets.items()}
+    avg_assets = {city:asset[0]/asset[1] for (city, asset) in city_assets.items()}
```
집계함수(sum, len) 없이 평균값 도출하도록 변경

eunjipark123 commented on tests/homework/xml_hw.py:
```diff
@@ -14,13 +14,10 @@ def find_poor_and_asset() -> Tuple:
-    return min_asset
+    
+    # 최소값 출력
+    min_asset = min(data, key=lambda x: int(x['est_asset_dollar']))
```
1. 인덴트 수정
2. 중간 불필요한 과정 생략

eunjipark123 commented on tests/homework/xml_hw.py:
```diff
@@ -31,20 +28,23 @@ def find_top3_poorest_city() -> list:
-        sorted_lists = sorted(avg_assets.items(), key=lambda x: x[1], reverse=False)
-        bottom3 = [city[0] for city in sorted_lists[:3]]
+    # 도시 별 평균 자산액 계산
+    city_assets = {}
```
인덴트 수정

eunjipark123 commented on tests/homework/xml_hw.py:
```diff
@@ -31,20 +28,23 @@ def find_top3_poorest_city() -> list:
+            city_assets[city][0] += asset 
+            city_assets[city][1] += 1
+
+    avg_assets = {city:asset[0]/asset[1] for (city, asset) in city_assets.items()}
```
sum, len 함수 없이 바로 sum, cnt값 계산하도록 변경

Moon-edu commented:
> 아핫 이해했습니다. 로직을 여러 번 생각해보면서 효율적인 방법을 찾아봐야겠네요!! 제가 여러 번 헤메서 많이 답답하시죠 🥲🥲 그래도 계속 봐주시고 좋은 방향으로 나아갈 수 있게 피드백 해주셔서 정말 감사드립니다! ㅎㅎ

코드리뷰라는 것이 원래 핑퐁 게임입니다. 리뷰어가 이슈 제기하고 리뷰이가 수정하고, 모두가 "이쯤에서면 문제 없겠다"라는 지점이 올 때 Approve하고 Merge하는 과정이고 꽤나 험난한 과정이기도 합니다.  
어떠한 미안한 감정을 가질 필요가 전혀 없으며 단지 개선해 나가고 있는 과정일뿐으로, 이를 통해 계속 발전해나가면 됩니다.  
실무에서도 마찬가지입니다. 저도 입사한 첫 해 수많은 코멘트 수정하느라 끙끙댔었는데 그 과정에서 더 좋은 알고리즘과 코딩스타일, 동료들과 의사소통을 배웠던 것 같습니다. 저와의 코드리뷰 과정에서 코드를 짜는 방법뿐만이 아니라, 이런 코드리뷰 과정 그 자체, 끈임없이 자신의 코드를 의심하는 습관, 배우는 법을 배우는 것 등을 가져가셨으면 합니다.

Moon-edu commented:
전반적으로 동작하는 코드들이고, 크겍 문제가 없는 코드라고 생각되기때문에, 더 많은 시간을 들여서 코드를 수정해도 괜찮고, 이쯤에서 마무리해도 충분히 괜찮다고 생각하기때문에 여기서 PR을 닫고 다음으로 넘어가고 싶다면 말씀주세요. 리뷰어로서 이 정도면 Approve를 할 것 같구요, 코드 변경을 계속하시면 그것 역시 괜찮습니다. 제가 계속 확인해서 피드백 드리겠습니다.

eunjipark123 committed:
오타 수정

변수 오타 수정 (sum_essets > sum_assets)

eunjipark123 committed:
오타수정2

eunjipark123 committed:
수정 제출

1. max나 len, sum 등 비싼(?) 함수 대신 반복문으로 변경했습니다.
2. 데이터 불러오고 sum, cnt 계산 -> json데이터를 불러올 때 바로 변수에 저장하는 방식으로 변경했습니다.

eunjipark123 commented:
수정 제출한 부분이 더 효율적인지 한번 더 체크 부탁드립니다! 
강사님의 꼼꼼한 리뷰 덕분에 정말 많이 배우고 있습니다. 감사합니다!!!

Moon-edu commented on tests/homework/xml_hw.py:
```diff
@@ -4,19 +4,57 @@
+    sorted_lists = sorted(avg_assets.items(), key=lambda x: x[1], reverse=False)[:3]
+    bottom3 = [city[0] for city in sorted_lists]
+
+    return bottom3
```
사실 가장 최적화된 Top/Bottom K 찾기 알고리즘은 따로 있습니다. Heap 이라는 자료구조를 이용하면 공간복잡도 O(1), 시간복잡도 O(n)으로 뽑아낼 수 있습니다.(Top 3이라고 고정된 경우, Top K일 경우 O(nlogk)) 다른 분이 작성한 [코드 링크](https://github.com/Moon-edu/de-course-week4/pull/7/files#diff-6b497085868fc154b4fdc0b30d4a3cc52e2d67a95617c6a2f0d60a8d3fa6534eL21) 드립니다. 코드만 보고 이해하기 어렵다면 Chat GPT에 `how to extract top 3 elements from list with heap` 물어보면 잘 설명해줄 겁니다. 
혹은 조금 단순하긴 하지만 그냥 if else로 top3개만 유지하면 O(n)의 시간 복잡도, O(1)의 공간복잡도로 표현할 수 있습니다.(코드를 ChatGPT 답변 그대로 복사해왔습니다)
```
def extract_top3_linear(input_list):
    top3 = [float('-inf')] * 3

    for num in input_list:
        if num > top3[0]:
            top3[2] = top3[1]
            top3[1] = top3[0]
            top3[0] = num
        elif num > top3[1]:
            top3[2] = top3[1]
            top3[1] = num
        elif num > top3[2]:
            top3[2] = num

    return top3
```

eunjipark123 replied:
부끄럽게도 heap 자료구조는 처음 들어보았네요!! 공부해보고 한번 다시 적용해 보겠습니다. 감사합니다 !

Moon-edu commented on tests/homework/xml_hw.py:
```diff
@@ -4,19 +4,57 @@
+        data = xmltodict.parse(xml_file.read())['data']['asset']
+
+    
+    # 최소값 출력 (min 함수 안쓰고)
```
이전 로직이 잘 기억나지 않는데 min 함수를 쓴다는게 퍼포먼스를 떨어뜨린다는 의미는 아니었습니다. 
오해없으시길 바라겠습니다.  
아마 for문에서 min을 쓰셨거나 뭔가 반복문을 한 후 min을 하여서 두 번 iteration을 했거나였을 것으로 추측됩니다.

eunjipark123 replied:
min, max, sum 같은 함수가 비싼 함수(?) 이라는 것이 기억났고, 리스트를 전체 돌려서 값을 계산하는 거라 시간이나 공간 복잡도가 오래 걸린다고 생각했었어요!.. 요럴 땐 써도 되는군요! ㅎㅎ

Moon-edu commented on tests/homework/json_hw.py:
```diff
@@ -4,9 +4,24 @@
+        # 불러올 때 조건 지정하기
+        for row in json.load(json_file):
+            asset_sum += int(row['est_asset_dollar'])
+            asset_cnt += 1
```
요거는 사실 파일을 다 읽어온 후 파일을 연채로 반복문을 돌리는 건데, 이전에 드렸던 피드백과 동일하게 적용되는 문제입니다.
```python
with open("hw_data/assets.json", "r") as json_file:
    data = json.load(json_file):

for row in data:
    asset_sum += int(row['est_asset_dollar'])
    asset_cnt += 1
```
이렇게 되어야 파일을 다 읽어오면 바로 닫죠. 주석에 `불러올 때 조건 지정하기`라고 써놓으셨는데 혹시 의미가 파일을 한 줄 한 줄 읽어오는 걸로 생각하시는 것 같은데 그렇지 않습니다. `load`를 호출한 순간 파일 전체를 읽어와서 리스트로 바인딩합니다.

eunjipark123 replied:
맞습니다..! 불러오면서 for문을 돌린다고 생각해서 위의 코드로 작성하였습니다. 이미 어차피 파일 전체를 읽어오는 거라면 밖에 꺼내둬도 괜찮겠네요! ㅎㅎ 감사합니다!

Moon-edu commented on tests/homework/csv_hw.py:
```diff
@@ -4,19 +4,55 @@
+
+    # 정렬 후 top3 도출
+    sorted_lists = sorted(avg_assets.items(), key=lambda x:x[1], reverse=True)[:3]
+    top3 = [(city[0]) for city in sorted_lists]
```
top3는 문자열 리스트인데 `(city[0])`때문에 마치 튜플 리스트인 것처럼 혼란을 주네요. 파이썬에서 괄호는 튜플을 만들때 쓰기때문에 이 코드는 좋지 않아보입니다.

eunjipark123 replied:
엇.. 그러게요! 왜 괄호를 쳐놓았을까요?!!

Moon-edu commented:
잘하셨고 PR은 여기서 닫도록 하겠습니다. Heap 관련된 부분은 개인적으로 공부하시면 도움될거에요, 알고리즘 문제에서 한 번씩 써먹을일 있는 자료구조거든요

eunjipark123 commented:
앗..ㅠㅠ 주중에 일이 있어서 팔로우업이 늦었습니다! 위에 말씀해주신 피드백 반영해보고, Heap 자료구조도 공부해보겠습니다. 감사합니다 :)
